flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -0,0 +1,235 @@
+            List<RequestEntryT> requestEntries, ResultFuture<RequestEntryT> requestResult);
+
+    /**
+     * Buffer to hold request entries that should be persisted into the destination.
+     *
+     * <p>A request entry contain all relevant details to make a call to the destination. Eg, for
+     * Kinesis Data Streams a request entry contains the payload and partition key.
+     *
+     * <p>It seems more natural to buffer InputT, ie, the events that should be persisted, rather
+     * than RequestEntryT. However, in practice, the response of a failed request call can make it
+     * very hard, if not impossible, to reconstruct the original event. It is much easier, to just
+     * construct a new (retry) request entry from the response and add that back to the queue for
+     * later retry.
+     */
+    private final Deque<RequestEntryT> bufferedRequestEntries = new ArrayDeque<>();
+
+    /**
+     * Tracks all pending async calls that have been executed since the last checkpoint. Calls that
+     * completed (successfully or unsuccessfully) are automatically decrementing the counter. Any
+     * request entry that was not successfully persisted needs to be handled and retried by the
+     * logic in {@code submitRequestsToApi}.
+     *
+     * <p>There is a limit on the number of concurrent (async) requests that can be handled by the
+     * client library. This limit is enforced by checking the queue size before accepting a new
+     * element into the queue.
+     *
+     * <p>To complete a checkpoint, we need to make sure that no requests are in flight, as they may
+     * fail, which could then lead to data loss.
+     */
+    private int inFlightRequestsCount;
+
+    @Override
+    public void write(InputT element, Context context) throws IOException, InterruptedException {
+        while (bufferedRequestEntries.size() >= maxBufferedRequests) {
+            mailboxExecutor.yield();
+        }
+
+        bufferedRequestEntries.add(elementConverter.apply(element, context));
+
+        flushIfFull();
+    }
+
+    private void flushIfFull() throws InterruptedException, IOException {
+        while (bufferedRequestEntries.size() >= maxBatchSize) {
+            flush();
+        }
+    }
+
+    /**
+     * Persists buffered RequestsEntries into the destination by invoking {@code
+     * submitRequestEntries} with batches according to the user specified buffering hints.
+     *
+     * <p>The method blocks if too many async requests are in flight.
+     */
+    private void flush() throws InterruptedException, IOException {
+        while (inFlightRequestsCount >= maxInFlightRequests) {
+            mailboxExecutor.yield();
+        }
+
+        List<RequestEntryT> batch = new ArrayList<>(maxBatchSize);
+
+        while (batch.size() <= maxBatchSize && !bufferedRequestEntries.isEmpty()) {
+            try {
+                batch.add(bufferedRequestEntries.remove());
+            } catch (NoSuchElementException e) {
```

Comment:
How is it possible that a `NoSuchElementException` can be thrown here? Since no concurrency is possible in this method and there are no other remove/augment methods called on `bufferedRequestEntries` anywhere else?

Reply:
It is... I checked with Steffen yesterday, and there used to be a different queue here that would use `O(n)` to do `.size()`. Now that we are using an `ArrayDeque` I've rearranged this section a bit.